### DISCUSSION TITLE
The command  "PATH=$PATH:/usr/local/bin:/opt/homebrew/bin NODE_PATH=`/usr/bin/nod /usr/bin/npm root -g`

### DISCUSSION THREAD

**Discussion by bashiirmahad** [spatie/browsershot#903] [General]
I want to know how to solve this. i tried many different environments and consistently showing this error Whatever solution and suggestion I try, it keeps showing the same. Right now I am using Deck Env on Windows; also, this is the same with Ubuntu.

On normal PC: 
![cmd](https://github.com/user-attachments/assets/845db037-697f-4c6b-8b07-40b2eab45a48)

On deck env:
![deck cmd](https://github.com/user-attachments/assets/e893801f-1087-4b36-8757-b4c37251d5d3)

The command "PATH=$PATH:/usr/local/bin:/opt/homebrew/bin NODE_PATH=`/usr/bin/nod /usr/bin/npm root -g` /usr/bin/nod '/var/www/vendor/spatie/browsershot/src/../bin/browser.cjs' '{"url":"file:\/\/\/tmp\/1954879133-0070553001732521312\/index.html","action":"pdf","options":{"path":"example.pdf","args":[],"viewport":{"width":800,"height":600},"displayHeaderFooter":false}}'" failed.

Exit Code: 127(Command not found)

Working directory: /var/www/public

Output:
================

Error Output:
================
sh: /usr/bin/nod: not found
sh: /usr/bin/nod: not found

**Comment by nathan-de-pachtere** [comment]
Same issue after upgrading from 4.1.0

**Comment by ryanmortier** [comment]
Same issue after upgrading to v4.

**Comment by lucasdcrk** [comment]
Same issue using Laravel Herd with spatie/laravel-pdf

**Comment by ryanmortier** [comment]
I solved my issue by just installing puppeteer locally (rather than globally) `npm install puppeteer` then making sure `npm install` is run on my production deploy script.

**Comment by paulhennell** [comment]
Had this same issue where it errored out like above with `sh: node: command not found ` indicating it's struggling to find the right path to run node/npm.

As this is the top find when you google the error, figured I'd note the real answer is in the package documentation under [Custom Node and NPM binaries](https://spatie.be/docs/browsershot/v4/requirements#content-custom-node-and-npm-binaries). Locate the Node/NPM paths and pass those to Browsershot so it knows where they are. (Note - if the paths are different in your local to production don't hardcode it! 😉)

And if you're using `spatie/laravel-pdf` you need to pass the paths into the _browsershot_ instance not the PDF facade - see the documentation [on the withBrowsershot() method](https://spatie.be/docs/laravel-pdf/v1/advanced-usage/customizing-browsershot) for how that works).

Once I set that up my PDFs started working as expected! 🎉

**Comment by migsAV** [comment]
The package allows you to set the `includePath`, which is the path to where it looks for `node`.

You will need to explicitly provide the path yourself when you are not using the homebrew node package, and this goes for when using it, as @paulhennell has mentioned in the `spatie/laravel-pdf` package.

## Browsershot
[spatie/browsershot](https://spatie.be/docs/browsershot/v4/usage/introduction)

**<ins>Original</ins>**
```
Browsershot::url('https://example.com')
->save('example.pdf');
```

**<ins>Modifified</ins>**
```
Browsershot::url('https://example.com')
->setIncludePath('$PATH:/usr/local/bin:/full_home_path/Herd/config/nvm/versions/node/version_of_node/bin/') // Custom path here
->save('example.pdf');
```

## Laravel-pdf
[spatie/laravel-pdf](https://spatie.be/docs/laravel-pdf/v1/advanced-usage/customizing-browsershot)

**<ins>Original</ins>**
```
Pdf::view('test')
    ->withBrowsershot(function (Browsershot $browsershot) {
        $browsershot->scale(0.5);
    })
    ->save($this->targetPath);
```

**<ins>Modifified</ins>**
```
 Pdf::view('test')
    ->withBrowsershot(function (Browsershot $browsershot) {
        $browsershot
        ->setIncludePath('$PATH:/usr/local/bin:/full_home_path/Herd/config/nvm/versions/node/version_of_node/bin/') // Custom path here
        ->scale(0.5);
    })
    ->save($this->targetPath);
```
